src/Nav.js

```diff
@@ -77,19 +143,80 @@ class Nav extends React.Component {
     return child.props.active;
   }
 
-  renderNavItem(child, index) {
-    return cloneElement(
-      child,
-      {
-        role: this.props.bsStyle === 'tabs' ? 'tab' : null,
-        active: this.getChildActiveProp(child),
-        activeKey: this.props.activeKey,
-        activeHref: this.props.activeHref,
-        onSelect: createChainedFunction(child.props.onSelect, this.props.onSelect),
-        key: child.key ? child.key : index,
-        navItem: true
-      }
+  getTabProps(child, idx, isActive, onSelect) {
+    let {
+        linkId, 'aria-controls': controls
+      , onKeyDown, eventKey, tabIndex = 0 } = child.props;
+
+    let context = this.getContext('$bs_tabcontainer');
+
+    if (!context.getId) {
+      return {};
+    }
+
+    warning(!(linkId || controls),
+      'In the context of a TabContainer, NavItems are given generated `linkId` and `aria-controls` ' +
+      'attributes for the sake of proper component accessibility. Any provided ones will be ignored. ' +
+      'To control these attributes directly provide a `generateChildId` prop to the parent TabContainer.'
     );
+
+    linkId = context.getId(eventKey, TAB) || null;
+    controls = context.getId(eventKey, PANE) || null;
+    onSelect = chain(onSelect, context.onSelect);
+
+    return {
+      onSelect,
+      linkId,
+      role: 'tab',
+      'aria-controls': controls,
+      tabIndex: isActive ? tabIndex : -1,
+      onKeyDown: chain(
+        this.handleTabKeyDown.bind(this, onSelect || (()=>{})),
+        onKeyDown
+      )
+    };
+  }
+
+  handleTabKeyDown(onSelect, event) {
```

Comment:
So this is a problem with moving the context stuff down to a TabNavItem, the focus behavior needs to be able to trigger `onSelect`, either this here _still_ needs to listen to tab container context, or some sort of ugly calling down to the child method.